cmd/generateproxy/interceptor.go

```diff
@@ -0,0 +1,374 @@
+	}
+
+	// If this type is a map then walk the underlying type and then make a note we need to encode maps with values of this type
+	if isMap(typ) {
+		result := walk(desired, elemType(typ))
+		record := records[typeName(typ)]
+		record.Map = true
+		return result
+	}
+
+	typeName := typeName(typ)
+
+	// If we've walked this type before, return the previous result
+	if record, ok := records[typeName]; ok {
+		return record.Matches
+	}
+
+	record := TypeRecord{}
+	records[typeName] = &record
+
+	for _, meth := range typeutil.IntuitiveMethodSet(elemType(typ), nil) {
+		// Ignore non-exported methods
+		if !meth.Obj().Exported() {
+			continue
+		}
+
+		methodName := meth.Obj().Name()
+		// The protobuf types have a Get.. method for every protobuf they refer to
+		// We walk only these methods to avoid cycles or other nasty issues
+		if !strings.HasPrefix(methodName, "Get") {
+			continue
+		}
+
+		sig := meth.Obj().Type().(*types.Signature)
+		// All the Get... methods return the relevant protobuf as the first result
+		resultType := sig.Results().At(0).Type()
+
+		// Check if this method returns a Payload(s) or if it leads (eventually) to a Type which refers to a Payload(s)
+		if typeMatches(resultType, desired...) || walk(desired, resultType) {
+			record.Matches = true
+			record.Methods = append(record.Methods, methodName)
+		}
+	}
+
+	// Return whether this Type can (eventually) lead to Payload(s)
+	// This is used in the walk logic above so that our encoding handles intermediate Types between our Request/Response objects and Payload(s)
+	return record.Matches
+}
+
+func generateInterceptor(cfg config) {
+	conf := &packages.Config{Mode: packages.NeedImports | packages.NeedTypes | packages.NeedTypesInfo}
+	pkgs, err := packages.Load(conf, "go.temporal.io/api/workflowservice/v1")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	servicePkg := pkgs[0]
+
+	pkgs, err = packages.Load(conf, "go.temporal.io/api/common/v1")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	commonPkg := pkgs[0]
+	scope := commonPkg.Types.Scope()
+	payloadTypes := []types.Type{
+		scope.Lookup("Payloads").Type(),
+		scope.Lookup("Payload").Type(),
+	}
+
+	scope = servicePkg.Types.Scope()
+	// UnimplementedWorkflowServiceServer is auto-generated via our API package
+	// The methods on this type refer to all possible Request/Response types so we can use this to walk through all of our protobuf types
+	service := scope.Lookup("UnimplementedWorkflowServiceServer")
+	if _, ok := service.(*types.TypeName); ok {
+		for _, meth := range typeutil.IntuitiveMethodSet(service.Type(), nil) {
+			if !meth.Obj().Exported() {
+				continue
+			}
+
+			sig := meth.Obj().Type().(*types.Signature)
+			walk(payloadTypes, sig.Params().At(1).Type())
+			walk(payloadTypes, sig.Results().At(0).Type())
+		}
+	}
+
+	buf := &bytes.Buffer{}
+
+	err = interceptorTemplate.Execute(buf, pruneRecords(records))
+	if err != nil {
+		log.Fatal(err)
+	}
+	src, err := imports.Process(interceptorFile, buf.Bytes(), nil)
```

Comment:
Lets make sure we do a `format.Source` for this too

Reply:
`imports.Process` also formats.

Reply:
Oh. Hmm, maybe I misunderstood. I think it only formats the imports. Will put `format.Source` back in also.